File: app/models/raffle.rb
# == Schema Information
#
# Table name: raffles
#
#  id              :integer          not null, primary key
#  campaign_id     :integer
#  user_id         :integer
#  start_time      :datetime
#  end_time        :datetime
#  price_per_entry :decimal(5, 2)
#  status          :string(255)
#  created_at      :datetime
#  updated_at      :datetime
#

class Raffle < ActiveRecord::Base
  has_many          :raffle_prizes
  has_many          :prizes,
    :through => :raffle_prizes

  belongs_to :campaign

  delegate :donations, :donators,
    :to => :campaign

  scope :finalized, -> { where(:finalized => true) }
  scope :completed, -> { where(:finalized => false).where("end_time < ?", Time.now) }
  scope :in_progress, -> { where(:finalized => false).where("start_time < ? AND end_time > ?", Time.now, Time.now) }
  scope :pending, -> { where(:finalized => false).where("start_time > ?", Time.now) }

  def status
    status = "pending"
    if finalized?
      status = "finalized"
    elsif ended?
      status = "completed"
    elsif started?
      status = "in_progress"
    end
    status
  end

  def started?
    start_time.present? && start_time < Time.now
  end

  def ended?
    end_time.present? && end_time < Time.now
  end

  def valid_donations
    donations.where("donations.amount > ?", price_per_entry).where("donations.created_at > ? AND donations.created_at < ?", start_time, end_time)
  end


end

File: app/assets/javascripts/ng-app/factories/Raffle.js
angular.module("raffle")
  .factory("Raffle", ["$log", "raffleService",
    function($log, raffleService) {
      function Raffle(data) {
        if (typeof data !== "undefined") {
          this.id = data.id;
          this.campaign_id = data.campaign_id;
          this.user_id = data.user_id;
          this.start_time = new Date(data.start_time);
          this.end_time = new Date(data.end_time);
          this.price_per_entry = data.price_per_entry;
          this.status = data.status;
          this.created_at = data.created_at;
          this.updated_at = data.updated_at;
          this.prizes = data.prizes;
        } else {
          this.prizes = [];
        }
      }

      Raffle.prototype.save = function() {
        if (this.id) {
          raffleService.put(this).then(function(response) {
            $log.log(response);
          }, function(error) {
            $log.log(error);
          })
        } else {
          raffleService.post(this).then(function(response) {
            $log.log(response);
          }, function(error) {
            $log.log(error);
          })
        }
      };

      return Raffle;
    }
  ])
